File: Difficulty: Medium/Subarrays With At Most K Distinct Integers/subarrays-with-at-most-k-distinct-integers.java
import java.util.*;

class Solution {
    public int countAtMostK(int[] arr, int k) {
        int left = 0, res = 0;
        Map<Integer, Integer> freq = new HashMap<>();

        for (int right = 0; right < arr.length; right++) {
            freq.put(arr[right], freq.getOrDefault(arr[right], 0) + 1);

            if (freq.size() > k) {
                while (freq.size() > k) {
                    freq.put(arr[left], freq.get(arr[left]) - 1);
                    if (freq.get(arr[left]) == 0) {
                        freq.remove(arr[left]);
                    }
                    left++;
                }
            }

            res += right - left + 1; // Count subarrays ending at right
        }

        return res;
    }
}

File: Difficulty: Medium/Subarrays With At Most K Distinct Integers/subarrays-with-at-most-k-distinct-integers.py
from collections import defaultdict

class Solution:
    def countAtMostK(self, arr, k):
        n = len(arr)
        freq = defaultdict(int)
        l = 0
        distinct = 0
        ans = 0

        for r in range(n):
            # include arr[r] in window
            if freq[arr[r]] == 0:
                distinct += 1
            freq[arr[r]] += 1

            # shrink window until distinct <= k
            while distinct > k:
                freq[arr[l]] -= 1
                if freq[arr[l]] == 0:
                    distinct -= 1
                l += 1

            # all subarrays ending at r with start in [l..r] are valid
            ans += (r - l + 1)

        return ans
